src/string.rs

```diff
@@ -0,0 +1,129 @@
+use std::{
+    fmt::Display,
+    ops::{Deref, DerefMut},
+    str::{from_utf8_unchecked, from_utf8_unchecked_mut},
+};
+
+#[derive(Debug)]
+pub enum CowStr<'s> {
+    Owned(String),
+    Borrowed(&'s str),
+    Inlined(InlineStr),
+}
+
+impl<'s> Deref for CowStr<'s> {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            Self::Owned(ref s) => s,
+            Self::Borrowed(s) => s,
+            Self::Inlined(s) => s.deref(),
+        }
+    }
+}
+
+impl<'s> AsRef<str> for CowStr<'s> {
+    fn as_ref(&self) -> &str {
+        self.deref()
+    }
+}
+
+impl<'s> From<&'s str> for CowStr<'s> {
+    fn from(value: &'s str) -> Self {
+        CowStr::Borrowed(value)
+    }
+}
+
+impl<'s> From<String> for CowStr<'s> {
+    fn from(value: String) -> Self {
+        CowStr::Owned(value)
+    }
+}
+
+impl<'s> Clone for CowStr<'s> {
+    fn clone(&self) -> Self {
+        match self {
+            CowStr::Owned(s) => match InlineStr::try_from(&**s) {
+                Ok(inline) => CowStr::Inlined(inline),
+                Err(_) => CowStr::Owned(s.clone()),
+            },
+            CowStr::Borrowed(s) => CowStr::Borrowed(s),
+            CowStr::Inlined(s) => CowStr::Inlined(*s),
+        }
+    }
+}
+
+impl<'s> PartialEq for CowStr<'s> {
+    fn eq(&self, other: &Self) -> bool {
+        self.deref() == other.deref()
+    }
+}
+
+impl<'s> Eq for CowStr<'s> {}
+
+impl<'s> Display for CowStr<'s> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.deref())
+    }
+}
+
+impl<'s, 'a> FromIterator<&'a str> for CowStr<'s> {
+    fn from_iter<T: IntoIterator<Item = &'a str>>(iter: T) -> Self {
+        CowStr::Owned(FromIterator::from_iter(iter))
+    }
+}
+
+const MAX_INLINE_STR_LEN: usize = 3 * std::mem::size_of::<isize>() - 2;
```

Comment:
-2 is one byte for the length and one byte for the tag?

Reply:
I gotta admit, this is a remnant from pulldown_cmark which I pulled over before thinking any more of it. With further thought, it should really be

```rust
std::mem::size_of::<String> - 1;
```

to make it the same size as a `String`.

Reply:
The 2 is still needed for the tag, right?

Right now, it seems like CowStr is 32 bytes (1B tag + 7B padding + 24B String). So we should be able to have 4 words - 2B right? I.e. 1B tag + 1B len + 30B data.

Reply:
You are entirely right, I did not consider that we of course have more room because of the tag padding. Playground to prove that we indeed can use 30 bytes as our `MAX_INLINE_STR_LEN`:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e4df58ea66892efb6bbd444ff3fc2c8b

However, I have a question that I'm sure you know the answer too: Won't 32-bit architectures align by 4 bytes, thus only giving us room for a `MAX_INLINE_STR_LEN` of 26 bytes?

This low-level stuff is pretty new to me, gotta admit - but I'm very happy to learn!

Reply:
Yes, it will align to 4 bytes, because the fields are each 4 bytes instead. A String consists of a pointer to the data, the current length and the capacity. Each of these three needs to match the address space size so for 32-bit systems, they are each 32 bits long (1 word). So the size of the Owned variant will be 3 words + 1 word of tag+padding (`4 * sizeof(usize)`). To match the same size for the Inlined variant, we can use 4 words minus two bytes (length+tag) for the data. So e.g. 14 bytes for 32-bit, 6 bytes for 16-bit.

Reply:
`size_of::<String> - 1` should be possible, since the lenght does not need a full byte (if the string is inlined, it can't be larger than this size, e.g. no larger than 23 on most platforms), so it should be possible to fit the length and the tag in the same byte.

Reply:
@kaj Indeed, that is how the `smartstring` crate does it. However, since we want to use the inlined string through our `CowStr` type, we end up with 32 bytes either way (on 64-bit, that is.) We might as well use the room that comes from alignment.

Reply:
@kaj Yeah, it should technically be possible to share one byte for the tag and length, as the tag needs two bits and the length needs 5 bits. Then we could potentially gain another byte for the data.

But is it possible to make the compiler do this optimization? There is a NonZeroU8 type, but we need more than one discriminant, it would be convenient if there was a U5 type that only used the 5 least significant bits and allowed the compiler to use the rest for the tag.

Reply:
@hellux, would the extra work and code complexity even be worth it?

Reply:
If the discriminant has to be handled manually, instead of just hinting something to the compiler to let it optimize it, then I don't think so.